src/python/pants/option/arg_splitter.py

```diff
@@ -167,38 +158,37 @@ def assign_flag_to_scope(flg: str, default_scope: str) -> None:
         for flag in global_flags:
             assign_flag_to_scope(flag, GLOBAL_SCOPE)
         scope, flags = self._consume_scope()
-        while scope:
-            if not self._check_for_help_request(scope.lower()):
-                add_scope(scope)
-                if scope in self._known_goal_scopes:
-                    goals.add(scope)
-                else:
-                    unknown_scopes.append(scope)
-                for flag in flags:
-                    assign_flag_to_scope(flag, scope)
+        while scope and add_goal(scope):
+            for flag in flags:
+                assign_flag_to_scope(flag, scope)
             scope, flags = self._consume_scope()
 
-        while self._unconsumed_args and not self._at_double_dash():
-            arg = self._unconsumed_args.pop()
-            if arg.startswith("-"):
+        while self._unconsumed_args and not builtin_goal and not self._at_double_dash():
+            if self._at_flag():
+                arg = self._unconsumed_args.pop()
                 # We assume any args here are in global scope.
-                if not self._check_for_help_request(arg):
-                    assign_flag_to_scope(arg, GLOBAL_SCOPE)
-            elif self.likely_a_spec(arg):
-                specs.append(arg)
-            elif arg not in self._known_scopes:
-                unknown_scopes.append(arg)
+                assign_flag_to_scope(arg, GLOBAL_SCOPE)
+                continue
 
-        if self._at_double_dash():
+            arg = self._unconsumed_args.pop()
+            if self.likely_a_spec(arg):
+                specs.append(arg)
+            else:
+                add_goal(arg)
+
+        if not builtin_goal:
+            if unknown_scopes and "__unknown_goal" in self._known_goal_scopes:
+                builtin_goal = "__unknown_goal"
```

Comment:
It'd be good to de-magic this value by using a constant to reduce the risk of us changing it and breaking things.